api/service.go

```diff
@@ -524,6 +528,158 @@ func (s *service) RemoveAppToken(ctx context.Context, req *pb.RemoveAppTokenRequ
 	return &pb.RemoveAppTokenReply{}, nil
 }
 
+// AddFolder handles an add folder request.
+func (s *service) AddFolder(ctx context.Context, req *pb.AddFolderRequest) (*pb.AddFolderReply, error) {
+	log.Debugf("received add folder request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	proj, err := s.getProjectForScope(ctx, req.ProjectID, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	pth, err := s.ipfsClient.Unixfs().Add(
+		ctx,
+		ipfsfiles.NewMapDirectory(map[string]ipfsfiles.Node{}),
+		options.Unixfs.Pin(true))
+	if err != nil {
+		return nil, err
+	}
+
+	folder, err := s.collections.Folders.Create(ctx, pth, req.Name, req.Public, proj.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.AddFolderReply{
+		ID:   folder.ID,
+		Path: folder.Path,
+	}, nil
+}
+
+// GetFolder handles a get folder request.
+func (s *service) GetFolder(ctx context.Context, req *pb.GetFolderRequest) (*pb.GetFolderReply, error) {
+	log.Debugf("received get folder request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	folder, err := s.getFolderWithScope(ctx, req.Name, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := s.ipfsClient.Unixfs().Get(ctx, path.New(folder.Path))
+	if err != nil {
+		return nil, err
+	}
+	defer node.Close()
+
+	return folderToPbFolder(folder, node)
+}
+
+func folderToPbFolder(folder *c.Folder, node ipfsfiles.Node) (*pb.GetFolderReply, error) {
+	entries, err := dirToPbSlice(node)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.GetFolderReply{
+		ID:        folder.ID,
+		Path:      folder.Path,
+		Name:      folder.Name,
+		Public:    folder.Public,
+		ProjectID: folder.ProjectID,
+		Created:   folder.Created,
+		Entries:   entries,
+	}, nil
+}
+
+func dirToPbSlice(node ipfsfiles.Node) (entries []*pb.GetFolderReply_Entry, err error) {
+	err = ipfsfiles.Walk(node, func(path string, n ipfsfiles.Node) error {
```

Comment:
So `node` has the real bytes data of files in the folder?

Curious since walking seems like a function that doesn't use some API to make this incremental, so I guess is all the data? (that means... GBs if the folder is big? or is a light structure?)

Reply:
Bytes aren't included—the node will be of type _File_ or _Directory_, both of which are `io.Reader`.

Reply:
Uhm, cool. I wonder how that is resolved. Maybe the `ipfsfiles.Node` hold a reference to the `ipfshttpclient` (or DagResolver) that was used to fetch them? Sounds like magic.

Reply:
Must be something like that... I just embraced the magic! Wild though.